tests/flava/test_modeling_flava.py

```diff
@@ -0,0 +1,1234 @@
+            )
+            self.parent.assertEqual(
+                result.contrastive_logits_per_text.shape,
+                (self.text_model_tester.batch_size, self.image_model_tester.batch_size),
+            )
+
+            for item in [
+                result.loss_info.global_contrastive,
+                result.loss_info.itm,
+                result.loss_info.mmm_text,
+                result.loss_info.mmm_image,
+            ]:
+                self.parent.assertEqual(item.dim(), 0)
+
+            for item in [result.loss_info.mim, result.loss_info.mlm]:
+                self.parent.assertIsNone(item)
+
+        else:
+            self.parent.assertIsNone(result.multimodal_masked_embeddings)
+            for item in [
+                result.loss_info.global_contrastive,
+                result.loss_info.itm,
+                result.loss_info.mmm_text,
+                result.loss_info.mmm_image,
+            ]:
+                self.parent.assertIsNone(item)
+
+        self.parent.assertIsNone(result.multimodal_embeddings)
+
+
+@require_torch
+class FLAVAForPreTrainingTest(FLAVAModelTest):
+    all_model_classes = (FLAVAForPreTraining,) if is_torch_available() else ()
+    class_for_tester = FLAVAForPreTrainingTester
+    test_torchscript = False
+
+
+# We will verify our results on an image of cute cats
+def prepare_img():
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    im = Image.open(requests.get(url, stream=True).raw)
+    return im
+
+
+@require_vision
+@require_torch
+class FLAVAModelIntegrationTest(unittest.TestCase):
+    @slow
+    def test_inference(self):
+        model_name = "aps/flava-full"
+        model = FLAVAModel.from_pretrained(model_name).to(torch_device)
+        processor = FLAVAProcessor.from_pretrained(model_name)
+
+        image = prepare_img()
+        inputs = processor(
+            text=["a photo of a cat", "a photo of a dog"],
+            images=[image, image],
+            padding="max_length",
+            max_length=77,
+            return_tensors="pt",
+            return_masks=False,
+        ).to(torch_device)
+
+        # forward pass
+        with torch.no_grad():
+            outputs = model(**inputs, return_dict=True)
+
+        # verify the embeddings
+        self.assertAlmostEqual(outputs.image_embeddings.sum().item(), -1352.53540, places=4)
+        self.assertAlmostEqual(outputs.text_embeddings.sum().item(), -198.98225, places=4)
+        self.assertAlmostEqual(outputs.multimodal_embeddings.sum().item(), -3988.51367, places=4)
+
+
+@require_vision
+@require_torch
+class FLAVAForPreTrainingIntegrationTest(unittest.TestCase):
+    @slow
+    def test_inference(self):
+        model_name = "aps/flava-full"
+        codebook_name = "aps/flava-codebook"
+        model = FLAVAForPreTraining.from_pretrained(model_name).to(torch_device)
+        codebook = FLAVACodebook.from_pretrained(codebook_name).to(torch_device)
```

Comment:
Hmm, don't really like that `flava-full` and `flava-codebook` are in two different repositories. All of our models usually work with one and only one modeling file. Can't we merge the `FlavaCodebook` a model into the `FlavaForPreTraining` class? Don't like that inference is dependent on two different repos

Reply:
I would ideally like to avoid that because codebook is somewhat independent of FLAVA itself as it is mostly just custom DALLE encoder. A different codebook can be potentially used with flava. If I join them into one, I will lose that benefit. wdyt?

Reply:
I see, I guess we have kind of a composition model problem here (maybe similar to Blenderbot2 or RAG cc @sgugger )  
My opinion here is the following. The paper only used the DALLE encoder (no?) and that's what we expect people to use when reproducing the paper. In our experience few people expect things that have not been done in the paper to work out of the box. 

The way it's currently implemented doesn't really allow the user to be able to use another encoder out-of-the-box either. E.g. the user would have to code up a new encoder class which respects the logic of: 

```py
mim_labels = codebook.get_codebook_indices(**codebook_fe([image, image], return_tensors="pt").to(torch_device))
```

i.e. the model would also require a `get_codebook_indices(...)` function etc... This requires already quite some knowledge about the model in which case it's probably as easy to fork `modeling_flava.py` and replace:


`self.codebook = FlavaCodebook` inside the `ForPretrainingClass` with a different encoder (might even be easier to understand.

Reply:
So my opinion here is that we should put `FlavaCodebook` inside the modeling classes that are needed so that we only have a single `pytorch_model.bin` file and which makes it easier to reproduce / reuse the default Flava as described in the paper

Reply:
In general "self-containment" is more important then "flexibility of modularization" for Transformers in my opinion.

Reply:
If we would have multiple pretrained checkpoints for different codebook classes we could have added a `codebook_type=dalle` flag to the config that would have switched in the correct codebook class, but nevertheless I think we should put in inside the `FlavaForPreTraining` class.

Let's see what @sgugger thinks here

Reply:
> The way it's currently implemented doesn't really allow the user to be able to use another encoder out-of-the-box either. E.g. the user would have to code up a new encoder class which respects the logic of:

Not really, because that is just one interface on how to create `mim_labels`. What FLAVAPretrainedModel is concerned about is `mim_labels` not where they came from or how they were created? I would ideally like to avoid creating `mim_labels` inside the pretraining class if possible.